### PR TITLE
fix(security): align BLOCKED_ENV_KEYS with the spawn denylist

### LINKED ISSUE #18423: fix(security): BLOCKED_ENV_KEYS is missing loader hijack keys the spawn denylist blocks
## Summary

`BLOCKED_ENV_KEYS` (`packages/agent/src/config/blocked-env-keys.ts`) is the denylist for config and API environment writes. Its own header names its first category as *"Process-level code injection (NODE_OPTIONS, LD_PRELOAD, ...)"*, and `config-routes.ts:426-429` states what it is for:

> this loop catches process-level injection keys (NODE_OPTIONS, LD_PRELOAD, etc.) so they never reach `saveElizaConfig()` and **the persistence→restart RCE chain is closed**.

For that category it is a strict subset of core's `BLOCKED_SPAWN_ENV_KEYS`. Nine keys are blocked there and not here:

`LD_AUDIT` · `DYLD_FRAMEWORK_PATH` · `PYTHONPATH` · `PYTHONSTARTUP` · `PYTHONHOME` · `PERL5OPT` · `PERL5LIB` · `RUBYOPT` · `RUBYLIB`

The list blocks `LD_PRELOAD` but not `LD_AUDIT`; `DYLD_INSERT_LIBRARIES` but not `DYLD_FRAMEWORK_PATH`; `NODE_OPTIONS` and `NODE_PATH` but no Python, Perl, or Ruby variable at all.

`LD_AUDIT` is the significant one. Core's own comment beside it reads *"ld.so audit hook: loads an arbitrary shared object into every spawned dynamically linked binary — same code-injection primitive as LD_PRELOAD."* The primitive this list already blocks under one name is unblocked under another.

## Why it is reachable

A key that survives this list is written into the parent process environment, and several spawn sites hand that environment to a child without filtering:

```
config-routes.ts:433   BLOCKED_ENV_KEYS.has(...)      the config/API write gate
config.ts:100          process.env[key] = value       survivors land in process.env
```

Then:

```
audio-redaction.ts:349   spawn(bin, args, { windowsHide: true })   no env -> inherits process.env
sandbox-engine.ts:142    spawn(binary, args, { stdio: [...] })     no env -> inherits process.env
workspace.ts:57          env: opts.env ?? process.env
shell-execution-router.ts:159   { ...process.env, ...sanitizeChildEnv(req.env) }
```

`sanitizeSpawnEnv` — the GHSA-54rx defense that does block these keys — filters only the **caller-supplied** env. It never inspects `process.env`. So the spawn-side mitigation cannot see anything that arrives through the config/API path, and `LD_AUDIT` or `DYLD_FRAMEWORK_PATH` set this way is injected into every dynamically linked child listed above, not just an interpreter.

## Scope, stated precisely

- The `LD_AUDIT` / `DYLD_FRAMEWORK_PATH` pair needs no particular child. Any dynamically linked binary spawned from an inherited environment loads the attacker's shared object.
- The Python/Perl/Ruby keys only bite when a matching interpreter is spawned. Nothing in this repository spawns one today, and MCP servers are launched with `{...config.env, PATH}` rather than `process.env`, so they are unaffected. They are included for consistency with the spawn denylist rather than on their own reachability.
- No exploit is claimed. The write path is the API/config surface the existing guard already assumes an attacker may reach; `packages/docs/security.md` frames the same threat model.

## Proposed fix

Add the nine keys, grouped with a comment for each primitive, and record the invariant in the header: this list's code-injection category must remain a superset of `BLOCKED_SPAWN_ENV_KEYS`, because that list guards a caller-supplied child environment while this one guards writes that reach `process.env`.

Add a test that asserts the superset relation so the two lists cannot drift apart again, plus negative assertions that ordinary keys (`NODE_ENV`, `PYTHON_VERSION`, `LD_FLAGS`, `RUBY_VERSION`) still pass.

The keys should be listed literally rather than spread from core at module scope: `mcp-server-config.ts` records that the core barrel must not be captured at module-init, and a `new Set([...BLOCKED_SPAWN_ENV_KEYS])` would do exactly that.

## Also out of scope: the prefix list has no counterpart here at all

`BLOCKED_ENV_KEYS` is consumed only through `.has(key.toUpperCase())` — exact match, no prefix logic. Core's spawn policy additionally carries `BLOCKED_SPAWN_ENV_PREFIXES`:

```
NPM_CONFIG_ · PNPM_ · YARN_ · BUN_CONFIG_ · UV_ · PIP_ · PIPX_
PYX_ · DENO_ · DOCKER_ · PODMAN_ · BASH_FUNC_
```

None of these is enforced on the config/API path — grep for any of them in `packages/agent/src` returns nothing. `BASH_FUNC_*` is the Shellshock prefix, and it reaches `process.env` and every inheriting child by the same route described above.

This is deliberately not fixed here. Closing it means replacing `Set.has()` with a predicate at ten call sites across three files, and `BLOCKED_ENV_KEYS` is also exported as a `Set<string>` through `ConfigRouteContext`, so the Set has to remain regardless. That is a different defect class with a different blast radius and belongs in its own change.

### PR DESCRIPTION
**Pull request by sailorpepe**
# Relates to

Closes #18423

Definition of Done: full standard in [`CONTRIBUTING.md`](../CONTRIBUTING.md).

- [x] This PR targets `develop` and is rebased onto the latest `origin/develop` with zero conflicts.
- [x] `bun install` was run after sync; focused suites, typecheck, and Biome recorded below.
- [x] A reviewer can confirm the change works without reading the code, from the evidence attached below.

# Contribution provenance

- AI assistance: `yes`

- Model(s) used: `anthropic/claude-opus-5`

- Agent tooling: `claude-code`

- Skill path: `N/A - no contribution skill used`

- Provenance status: `self-reported`

# Sync with develop

- [x] Rebased onto the latest `origin/develop`; zero conflicts.
- [x] Focused suites pass post-sync; see **Testing** below.

# Risks

Low and additive. Nine entries are added to `BLOCKED_ENV_KEYS`; none is removed, no control flow changes, and every consumer reads the same `Set` through `.has(key.toUpperCase())`.

Over-blocking was the failure mode I checked hardest for. Nothing in the repository — source, config, fixture, example, or documentation — sets any of the nine keys, so there is no known compatibility surface. The plugin config path (`server-helpers-plugin.ts:186`) checks a per-plugin `allowedParamKeys` allowlist *before* consulting this list, so the additions are defense-in-depth there rather than a new rejection surface. Four negative controls (`NODE_ENV`, `PYTHON_VERSION`, `LD_FLAGS`, `RUBY_VERSION`) are asserted to still pass.

# Background

## What does this PR do?

`BLOCKED_ENV_KEYS` guards config and API environment writes, and its header names its first category as process-level code injection. For that category it was a strict subset of core's `BLOCKED_SPAWN_ENV_KEYS`: it blocked `LD_PRELOAD` but not `LD_AUDIT`, `DYLD_INSERT_LIBRARIES` but not `DYLD_FRAMEWORK_PATH`, and `NODE_OPTIONS` but no Python, Perl, or Ruby variable at all.

`LD_AUDIT` is the significant one — core's own comment beside it reads *"ld.so audit hook: loads an arbitrary shared object into every spawned dynamically linked binary — same code-injection primitive as LD_PRELOAD."* The primitive already blocked under one name was unblocked under another.

A key that survives this list is persisted to `eliza.json` and applied to `process.env` on the next config load (`config.ts:100`), and several spawn sites hand that environment to a child without filtering:

```
audio-redaction.ts:349   spawn(bin, args, { windowsHide: true })   no env -> inherits
sandbox-engine.ts:142    spawn(binary, args, { stdio: [...] })     no env -> inherits
workspace.ts:57          env: opts.env ?? process.env
shell-execution-router.ts:159   { ...process.env, ...sanitizeChildEnv(req.env) }
```

`sanitizeSpawnEnv` filters only a caller-supplied env and never inspects `process.env`, so the spawn-side mitigation cannot see anything arriving through the config path. `config-routes.ts:426-430` describes this route as the one it exists to close.

This adds the nine keys grouped by primitive and records the invariant in the header. A test asserts the superset relation so the two lists cannot drift apart again.

The keys are listed literally rather than spread from core at module scope: `mcp-server-config.ts` records that the core barrel must not be captured at module-init, and `new Set([...BLOCKED_SPAWN_ENV_KEYS])` would do exactly that.

## What kind of change is this?

Bug fixes (non-breaking change which fixes an issue)

# Documentation changes needed?

My changes do not require a change to the project documentation.

# Testing

## Where should a reviewer start?

`packages/agent/src/config/blocked-env-keys.ts` — the added entries and the header invariant. Then `blocked-env-keys.test.ts`, which asserts the superset relation, the nine keys individually, that `collectConfigEnvVars` drops them, and that ordinary keys still pass.

## Detailed testing steps

Reverting only the source file and re-running fails 3 of 6, so the assertions are bound to the change. The third failure is the reachability itself rather than a membership check — it shows the key surviving the collector that feeds `process.env`:

```
AssertionError: expected [ 'LD_AUDIT', …(8) ] to deeply equal []
AssertionError: expected false to be true
AssertionError: expected { LD_AUDIT: '/tmp/evil.so', …(5) }
              to deeply equal { SAFE_PUBLIC_FLAG: 'enabled', …(1) }
```

```
with the fix                                       6 passed (6)
source reverted, tests kept              3 failed | 3 passed (6)
packages/agent/src/config + model-config-routes    5 files, 52 passed
```

Biome clean on both changed files. `tsc --noEmit -p packages/agent/tsconfig.json` reports 26 errors both with and without this change — all pre-existing unbuilt-workspace-dependency errors (`@elizaos/cloud-routing`, `plugin-notes`), none in these files.

# Known gaps / failures

- No exploit is claimed. The loader behaviour is cited from ld.so/dyld semantics and core's own comment; it was **not reproduced here** — this host is macOS and `LD_AUDIT` is glibc. What the tests demonstrate is the reachability: the key survives the filter that feeds `process.env`.
- `LD_AUDIT` / `DYLD_FRAMEWORK_PATH` need no particular child. The Python/Perl/Ruby keys need a matching interpreter; no in-tree call site names one as its command, the shell router will spawn one if a user or agent runs it, and MCP servers receive `{...config.env, PATH}` rather than `process.env`, so they are unaffected. Those keys are included for consistency with the spawn denylist rather than on their own reachability.
- The persisted `config-env` store (`config.ts:251`) is applied to `process.env` without this filter by design. Every writer uses a hardcoded key constant (`first-time-setup.ts`, `remote-capability-routes.ts`) or migrates keys already present (`vault-bootstrap.ts`), so it is a closed set; it is noted here only so the asymmetry is not mistaken for an oversight.
- The superset test couples the two lists' release order: if #18417 lands first, its keys must be added here too or this test fails. That coupling is the point — it makes drift a build failure — but it is a real constraint and is one deletable test block if you would rather not carry it.
- **Overlap with #18473.** That PR derives `BLOCKED_ENV_KEYS` from `BLOCKED_SPAWN_ENV_KEYS` by spread, which covers all nine keys added here. If it lands first these nine become redundant and this PR should be closed; it is held open until then rather than leaving the gap to an unmerged PR.
- The prefix list has no counterpart on this path at all (`BASH_FUNC_`, `NPM_CONFIG_`, `DENO_`, …). That is a different defect class requiring a predicate at ten call sites, and is recorded in #18423 as a separate follow-up rather than bundled here.

# Evidence

- [x] N/A - no rendered UI; this changes a `packages/agent` denylist constant.

- [x] N/A - no rendered UI; behavior is asserted by the unit tests above.

- [x] N/A - the observable behavior is a set-membership check; the revert-and-fail run is the demonstration.

- [x] Full verification log attached below: the gap measured on clean develop, tests with the fix, the same tests with the source reverted, and the full config suite.
[evidence-blocked-env-keys.txt](https://github.com/user-attachments/files/30944756/evidence-blocked-env-keys.txt)

- [x] N/A - no frontend surface is touched.

- [x] N/A - no agent, prompt, provider, action, or model behavior changed.

- [x] N/A - no database, memory, wallet, scheduled-task, or generated-file output changed.

---

AI provider/model: Anthropic / claude-opus-5
Client / agent tooling: Claude Code
Contribution skill revision: N/A - no contribution skill used
Attribution status: self-reported

### CODE REVIEW

File: packages/agent/src/config/blocked-env-keys.test.ts [hunk 1]
```diff
@@ -21,6 +21,63 @@ describe("BLOCKED_ENV_KEYS", () => {
...
+    // list guards config/API writes, which reach process.env and are then
+    // inherited by every spawn site that passes no explicit env. A key blocked
+    // there but not here is reachable through that path.
+    const { BLOCKED_SPAWN_ENV_KEYS } = await import(
+      "@elizaos/core/security/spawn-env-policy"
+    );
+    const missing = [...BLOCKED_SPAWN_ENV_KEYS].filter(
+      (key) => !BLOCKED_ENV_KEYS.has(key),
+    );
+    expect(missing).toEqual([]);
```
Comment (tokenmaxxor) [inline packages/agent/src/config/blocked-env-keys.test.ts:35]: [P1] This invariant is already broken by the active companion PR #18417 in a valid, conflict-free merge order. On exact #18427 plus exact #18417's core policy, I measured 25 `BLOCKED_SPAWN_ENV_KEYS` absent from this agent set, so this assertion becomes red if #18427 merges first and #18417 lands afterward. The branches touch different packages, so Git does not force the rebase that would expose it. Please resolve the dependency before approval: wait for #18417's accepted set and update/rebase this list, or make the paired PR update both lists atomically. Do not copy its still-contested entries blindly; the blocker is the unresolved coordination and immediate `develop` regression.

**Review by tokenmaxxor — CHANGES_REQUESTED** [review]
Reviewed exact head `989a46b49b050a150eabe3b5af5bb71cedcc75ac` against `develop`, including the active paired PR #18417 at `f5eb834b908b7f0e3def25e6fd5f696341b766a2`.

Requesting changes for the merge-order regression attached inline. The nine additions are correct on the current base, but the new cross-package invariant is already incompatible with #18417's conflict-free core-only change. The dependency must be resolved before this exact head is safe to approve.

Independent verification in disposable, network-disabled trees:

- Raw #18427 diff audit — two regular TypeScript files; no manifest, lockfile, lifecycle script, binary, symlink, submodule, executable, or mode change
- Exact #18427 head on current `develop` core — PASS, 3/3 tests and 15 assertions: zero core keys missing, all nine new keys blocked/filtered, four near-key controls allowed
- Conflict-free paired-tree simulation using the exact #18417 `spawn-env-policy.ts` — PASS as a diagnostic, 3/3 tests and 15 assertions, and measured **25** core keys missing from #18427's agent list
- SHA-256 of the overlaid core policy matched #18417's exact blob input
- Current #18427 GitHub Security Advisory Gate — PASS

Because #18417 changes only `packages/core` while #18427 changes only `packages/agent`, Git has no file conflict to force a rebase. If #18427 merges first, #18417 can merge afterward and make this newly added superset test fail on `develop`. Please either wait for #18417's accepted key set and rebase/update this PR, or update the paired PR so the agent list and invariant move atomically. I am not recommending that unsettled/contested #18417 entries be copied blindly.

The full agent package/typecheck matrix was not independently reproduced; the PR reports baseline workspace type errors and GitHub exposed only the security/title gates. The focused exact-module and paired-branch probes are the applicable execution evidence for this finding.

Security routing: the finding concerns merge coordination between two already-public security PRs and discloses no new exploit path beyond #18423/#18416.

Evidence scope: backend denylist constants and tests. Focused source-boundary output is applicable. Screenshots, video, frontend logs, live-model trajectory, domain artifacts, and OCR are N/A because no UI, model, database, or generated artifact changes.

AI provider/model: openai / gpt-5.6-sol
Client / agent tooling: codex
Contribution skill revision: elizaOS/army@9259107132edeab02d9e47dbb7ce383721bada77:skills/contribute-to-eliza
Compute receipt: 0 project-attributed tokens (unavailable; device-signed, locally reported)
Attribution status: self-reported
— [codex-wsl-worker]

**Comment by sailorpepe** [comment]
You're right, and the merge-order hazard is a real defect in this PR rather than a coordination nit. Thank you for the paired-tree simulation — measuring 25 core keys missing is exactly the failure I should have modelled and didn't.

My PR text said the two "touch different regions" and would need "a trivial rebase." That was wrong in a specific way: because #18417 is `packages/core`-only and this is `packages/agent`-only, **Git has no shared file to conflict on**, so nothing forces that rebase. The superset test then breaks on `develop` after the second merge with no signal at merge time.

Worth contrasting with this PR and #18439, which both modify `blocked-env-keys.ts` and `blocked-env-keys.test.ts`. I checked with a three-way merge: the source file auto-merges cleanly, but the test file conflicts in four regions, so whichever lands second is forced to rebase. A loud conflict is safe; the #18417 pairing is dangerous precisely because it is silent, which is what you caught.

## Proposed fix: drop the superset test from this PR

Of the two paths you offered, I don't think either is right while #18417 is unsettled:

- Rebasing this onto #18417 chains an otherwise-clean change to a PR now blocked behind your private report.
- Copying #18417's 25 entries here is the blind copy you advised against, and I agree — that set is contested and one member is under private review.

So I would remove only this block from `blocked-env-keys.test.ts`:

```ts
it("is a superset of core's spawn denylist so the two cannot drift", async () => {
  const { BLOCKED_SPAWN_ENV_KEYS } = await import(
    "@elizaos/core/security/spawn-env-policy"
  );
  const missing = [...BLOCKED_SPAWN_ENV_KEYS].filter(
    (key) => !BLOCKED_ENV_KEYS.has(key),
  );
  expect(missing).toEqual([]);
});
```

No import change is needed — `BLOCKED_ENV_KEYS` is still referenced by the remaining assertions, and that block holds the file's only `@elizaos/core` import, so removing it leaves this PR with no cross-package reference at all.

What remains asserted is only what this PR changes:

- the canonical union used by API writes and startup env sync
- the nine loader/interpreter keys reachable through `process.env`
- that `collectConfigEnvVars` drops them during startup env collection
- that ordinary keys which only look loader-adjacent are still allowed
- the canonical secret keys

None of those reference a core symbol, so they are stable in any merge order.

The invariant is still correct and worth having; it just cannot be asserted from whichever package merges first. It belongs in the PR that lands last, once the core key set is accepted. For what it's worth, #18439's analogous prefix-family invariant is not exposed to this — #18417 adds keys only and does not touch `BLOCKED_SPAWN_ENV_PREFIXES`.

Happy to push that removal here, or to add the invariant to #18417 instead if you'd rather it move with the core change. Which would you prefer?

**Comment by NubsCarson** [comment]
**🛠️ shipwright · review — ✅ APPROVE (adversarially verified, 1 candidate refuted).**

Correct, additive, net-positive: the exact-match superset relation now genuinely holds (all 27 core spawn keys present, 9 additions match core spellings verbatim), uppercased exact-match fails closed, and the LD_AUDIT-blocked-under-one-name-but-not-another gap it fixes was real. The `@elizaos/core/security/spawn-env-policy` import resolves through the `./security/*` export so the drift guard actually runs (not vacuous).

**One follow-up note (not blocking):** core's denylist has two halves — this aligns the KEYS half; the **PREFIXES** half (most notably `BASH_FUNC_`) stays reachable via config→`process.env`→unfiltered spawn inheritance. You defer it to #18423 — good, just flagging it pairs with #18439's boot-path nit as the remaining live route.

File: packages/agent/src/config/blocked-env-keys.test.ts [hunk 2]
```diff
@@ -21,6 +21,63 @@ describe("BLOCKED_ENV_KEYS", () => {
     expect(BLOCKED_ENV_KEYS.has("OPINION_PRIVATE_KEY")).toBe(true);
   });
 
+  it("is a superset of core's spawn denylist so the two cannot drift", async () => {
```
Comment (isfinne) [inline packages/agent/src/config/blocked-env-keys.test.ts:24]: This is the right shape for the guard, and worth noting that it is order-sensitive by design: it compares against whatever `BLOCKED_SPAWN_ENV_KEYS` holds at run time, so #18417's 25 new core keys will fail it until this list catches up. Rebase whichever of the two merges second.

Scope note for a future reader: this asserts over the exact-key set only. The prefix families (`BASH_FUNC_*`, `NPM_CONFIG_*`, `UV_*`, …) are covered by the sibling test added in #18439, not here.

**Review by isfinne — APPROVED** [review]
Approving — the mechanism is the right one. Making the config-write denylist a *tested* superset of core's spawn denylist turns "remember to update both" into a CI failure, and the header comment explaining why the keys are listed literally instead of spread from core (the `mcp-server-config.ts` module-init constraint) answers the obvious "why not just re-export it" question up front.

Two coordination notes, neither a defect in this PR.

**1. This will go red the moment #18417 lands.** #18417 adds 25 keys to `BLOCKED_SPAWN_ENV_KEYS` that this branch's `BLOCKED_ENV_KEYS` does not carry, so whichever of the two merges second needs a rebase first. Computed against the two PR heads (`989a46b` here, `f5eb834` there):

```
core exact keys @18417 head : 52
agent BLOCKED_ENV_KEYS here : 40
in core but not here (25):
  BASH_ENV  CLASSPATH  GCONV_PATH  GEM_HOME  GEM_PATH  GIT_EXTERNAL_DIFF
  GIT_SSH_COMMAND  GLOBIGNORE  HOSTALIASES  IFS  JAVA_TOOL_OPTIONS  LOCALDOMAIN
  NLSPATH  PERLIO_DEBUG  PS4  PYTHONINSPECT  PYTHONUSERBASE  PYTHONWARNINGS
  RES_OPTIONS  RUBYSHELL  SHELLOPTS  TERMCAP  TERMINFO  TERMINFO_DIRS
  _JAVA_OPTIONS
```

That is the guard doing its job rather than a bug — worth calling out only so it is not read as an unrelated CI flake on the second merge.

**2. The superset test covers the exact-key set only**, not `BLOCKED_SPAWN_ENV_PREFIXES` — so `BASH_FUNC_*`, `NPM_CONFIG_*`, `UV_*`, `DENO_*` and the rest would still drift. I see #18439 adds `BLOCKED_ENV_KEY_PREFIXES` with the matching drift test, so this is already handled in the stack; noting it only so the gap between the two PRs is not mistaken for the finished state.

Verified while reading: `collectConfigEnvVars` upper-cases before the lookup (`env-vars.ts:122`, `:136`, `:220`) and every key added here is already upper-case, so the new entries are reachable through all three call sites.

**Comment by sailorpepe** [comment]
Flagging overlap for whoever sequences this: #18473 now derives the agent denylist from `BLOCKED_SPAWN_ENV_KEYS` directly, which would make these nine explicit additions redundant if it lands first. I checked the set — all nine keys added here are present in that PR's spawn key list at `d770f253`.

I'm happy to close this the moment #18473 merges. Leaving it open until then so the gap isn't left to an unmerged PR.

@tokenmaxxor — the merge-order hazard you raised is resolved either way: if #18473 lands, the set is derived rather than duplicated, so it cannot drift.
